### PR TITLE
Steps to Care - Notification Preferences and Workers Only

### PR DESCRIPTION
**Pull request by nateh777**
##### What does the change add or fix?

- Add the new Workers Only column to the data model and table. 
- Add a new person attribute for notification preferences. 
- Update the communication migration to include the proper URL's and SMS values. 
- Update the Job for the new notification preferences and sending SMS.
- Added a Lava shortcode via migration for note template/quick notes capability.
- Added migration for new security permissions added in https://github.com/KingdomFirst/RockBlocks/pull/85

---------

##### What does the change add or fix in a succinct statement that will be read by clients?

- Add the new Workers Only column to the data model and table. 
- Add a new person attribute for notification preferences. 
- Update the communication migration to include the proper URL's and SMS values. 
- Update the Job for the new notification preferences and sending SMS.

---------

##### Who reported, requested, or paid for the change?

Rock Springs

---------

##### Does this update or add options to the block UI?

No

---------

##### What files does it affect?

- rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs
- rocks.kfs.StepsToCare/Migrations/003_CreateCommunications.cs
- rocks.kfs.StepsToCare/Migrations/006_AddWorkersOnlyColumn.cs
- rocks.kfs.StepsToCare/Migrations/007_CreateAttribute.cs
- rocks.kfs.StepsToCare/Migrations/008_LavaShortcode.cs
- rocks.kfs.StepsToCare/Migrations/009_CreatePermissions.cs
- rocks.kfs.StepsToCare/Model/CareNeed.cs
- rocks.kfs.StepsToCare/Properties/AssemblyInfo.cs
- rocks.kfs.StepsToCare/SystemGuid/PersonAttribute.cs
- rocks.kfs.StepsToCare/rocks.kfs.StepsToCare.csproj

---------

##### Is it a breaking change for other versions/clients?

Requires Rock 12.4+ due to GetAttributeValue by Guid

### CODE REVIEW

File: rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs [hunk 1]
```diff
@@ -204,20 +216,50 @@ public virtual void Execute( IJobExecutionContext context )
...
-                            var errors = new List<string>();
-                            emailMessage.Send( out errors );
+                            var notificationType = assignee.PersonAlias.Person.GetAttributeValue( SystemGuid.PersonAttribute.NOTIFICATION.AsGuid() );
 
-                            if ( errors.Any() )
+                            if ( notificationType == null || notificationType == "Email" || notificationType == "Both" )
                             {
-                                errorCount += errors.Count;
-                                errorMessages.AddRange( errors );
+                                emailMessage.AddRecipient( new RockEmailMessageRecipient( assignee.PersonAlias.Person, mergeFields ) );
```
Comment (nateh777) [inline rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs:223]: Test if empty email does it generate an error. if not match the sms error for email.

File: rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs [hunk 2]
```diff
@@ -278,23 +353,46 @@ public virtual void Execute( IJobExecutionContext context )
                             continue;
                         }
                         var emailMessage = new RockEmailMessage( outstandingNeedsCommunication );
-                        //var smsMessage = new RockSMSMessage( outstandingNeedsCommunication );
+                        var smsMessage = new RockSMSMessage( outstandingNeedsCommunication );
                         //var pushMessage = new RockPushMessage( outstandingNeedsCommunication );
                         var recipients = new List<RockMessageRecipient>();
-                        var errors = new List<string>();
+                        var smsNumber = assigned.PersonAlias.Person.PhoneNumbers.GetFirstSmsNumber();
```
Comment (nateh777) [inline rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs:359]: Need the empty sms check in this for loop as well.

File: rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs [hunk 3]
```diff
@@ -305,12 +403,21 @@ public virtual void Execute( IJobExecutionContext context )
...
+                        else
+                        {
+                            assignedPersonSms++;
+                        }
                     }
                 }
 
             }
             context.Result = string.Format( "{0} emails sent", assignedPersonEmails );
+            context.Result += string.Format( "{0} SMS messages sent", assignedPersonSms );
```
Comment (nateh777) [inline rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs:420]: Add new line here
```suggestion
            context.Result += string.Format( "\n{0} SMS messages sent", assignedPersonSms );
```

File: rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs [hunk 4]
```diff
@@ -182,16 +184,26 @@ public virtual void Execute( IJobExecutionContext context )
                 linkedPages.Add( "CareDetail", detailPageRoute != null ? detailPageRoute.Route : "page/" + detailPage.Id );
                 linkedPages.Add( "CareDashboard", dashboardPageRoute != null ? dashboardPageRoute.Route : "page/" + dashboardPage.Id );
```
Comment (nateh777) [inline rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs:185]: Go ahead and add a root slash to be consistent matching the block setting.

File: rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs (hunk 1 above)
Comment (nateh777) [reply rocks.kfs.StepsToCare/Jobs/CareNeedAutomatedNotifications.cs]: It does generate an error, but not a very logical one:
> The parameter 'address' cannot be an empty string.
> Parameter name: address
So I added additional error handling.

**Review by guymaness2 — APPROVED** [review]

**Pull request merged**